connectors/tests/test_basic_rule.py

```diff
@@ -33,6 +99,219 @@
 }
 
 
+def basic_rule_one_policy_and_rule_uppercase():
+    basic_rule_uppercase = BASIC_RULE_ONE_JSON
+
+    basic_rule_uppercase["rule"] = basic_rule_uppercase["rule"].upper()
+    basic_rule_uppercase["policy"] = basic_rule_uppercase["policy"].upper()
+
+    return basic_rule_uppercase
+
+
+def contains_rule_one(basic_rules):
+    return any(is_rule_one(basic_rule) for basic_rule in basic_rules)
+
+
+def contains_rule_two(basic_rules):
+    return any(is_rule_two(basic_rule) for basic_rule in basic_rules)
+
+
+def contains_rule_three(basic_rules):
+    return any(is_rule_three(basic_rule) for basic_rule in basic_rules)
+
+
+def contains_default_rule(basic_rules):
+    return any(is_default_rule(basic_rule) for basic_rule in basic_rules)
+
+
+def is_rule_one(basic_rule):
+    return (
+        basic_rule.id_ == BASIC_RULE_ONE_ID
+        and basic_rule.order == BASIC_RULE_ONE_ORDER
+        and basic_rule.policy == Policy.from_string(BASIC_RULE_ONE_POLICY)
+        and basic_rule.field == BASIC_RULE_ONE_FIELD
+        and basic_rule.rule == Rule.from_string(BASIC_RULE_ONE_RULE)
+        and basic_rule.value == BASIC_RULE_ONE_VALUE
+    )
+
+
+def is_rule_two(basic_rule):
+    return (
+        basic_rule.id_ == BASIC_RULE_TWO_ID
+        and basic_rule.order == BASIC_RULE_TWO_ORDER
+        and basic_rule.policy == Policy.from_string(BASIC_RULE_TWO_POLICY)
+        and basic_rule.field == BASIC_RULE_TWO_FIELD
+        and basic_rule.rule == Rule.from_string(BASIC_RULE_TWO_RULE)
+        and basic_rule.value == BASIC_RULE_TWO_VALUE
+    )
+
+
+def is_rule_three(basic_rule):
+    return (
+        basic_rule.id_ == BASIC_RULE_THREE_ID
+        and basic_rule.order == BASIC_RULE_THREE_ORDER
+        and basic_rule.policy == Policy.from_string(BASIC_RULE_THREE_POLICY)
+        and basic_rule.field == BASIC_RULE_THREE_FIELD
+        and basic_rule.rule == Rule.from_string(BASIC_RULE_THREE_RULE)
+        and basic_rule.value == BASIC_RULE_THREE_VALUE
+    )
+
+
+def is_default_rule(basic_rule):
+    return (
+        basic_rule.id_ == BASIC_RULE_DEFAULT_ID
+        and basic_rule.order == BASIC_RULE_DEFAULT_ORDER
+        and basic_rule.policy == Policy.from_string(BASIC_RULE_DEFAULT_POLICY)
+        and basic_rule.field == BASIC_RULE_DEFAULT_FIELD
+        and basic_rule.rule == Rule.from_string(BASIC_RULE_DEFAULT_RULE)
+        and basic_rule.value == BASIC_RULE_DEFAULT_VALUE
+    )
+
+
+def test_include_lowercase_should_return_policy_include():
```

Comment:
I think this and below tests can be actually simplified - not sure what's the best way for that in Python, I found this as an example: https://docs.pytest.org/en/6.2.x/parametrize.html

So test can be parametrised, and you can pass your test cases as input and rewrite the test to something (pseudocode) like:

```python

test_case("include", Policy.INCLUDE)
test_case("INCLUDE", Policy.INCLUDE)
### etc. I'm not sure about actually correct syntax of this in Python
def test_policy_from_string_parsing_strings_correctly(parsed_string, expected_value)
    assert Policy.from_string(parsed_string) == expected_value
```

Reply:
Good point, that should probably work. I also thought about this, but was a bit worried that this could introduce flaky tests, if the parameter name matches a parameter from a test fixture (which I think gets injected automatically?). But from the docs it's probably not a problem as you're explicit about the parameter names in the `parametrize` decorator. I'll try out, let's see, if it works :)

Reply:
And it works, thank you for the hint!